tests/src/test/java/alluxio/client/fs/DirectAccessIntegrationTest.java

```diff
@@ -0,0 +1,102 @@
+import alluxio.client.file.FileSystemUtils;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.CreateFilePOptions;
+import alluxio.testutils.LocalAlluxioClusterResource;
+
+import com.google.common.io.ByteStreams;
+import com.google.common.io.Files;
+import org.apache.commons.io.IOUtils;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.io.File;
+
+/**
+ * Integration tests for path level configurations.
+ */
+public class DirectAccessIntegrationTest {
+  private static final byte[] TEST_BYTES = "TestBytes".getBytes();
+  private static final int USER_QUOTA_UNIT_BYTES = 1000;
+  private static final String DIRECT_DIR = "/mnt/direct/";
+  private static final String NON_DIRECT_DIR = "/mnt/non_direct/";
+
+  @Rule
+  public LocalAlluxioClusterResource mLocalAlluxioClusterResource =
+      new LocalAlluxioClusterResource.Builder()
+          .setProperty(PropertyKey.USER_FILE_BUFFER_BYTES, USER_QUOTA_UNIT_BYTES)
+          .setProperty(PropertyKey.USER_FILE_DIRECT_ACCESS, DIRECT_DIR)
+          .build();
+  private FileSystem mFileSystem;
+
+  private final String mLocalUfsPath = Files.createTempDir().getAbsolutePath();
+
+  @Before
+  public void before() throws Exception {
+    mFileSystem = mLocalAlluxioClusterResource.get().getClient();
+    mFileSystem.mount(new AlluxioURI("/mnt/"), new AlluxioURI(mLocalUfsPath));
+  }
+
+  @Test
+  public void writeDirect() throws Exception {
+    final int n = 3;
+    for (int i = 0; i < n; i++) {
+      AlluxioURI uri = new AlluxioURI(DIRECT_DIR + i);
+      try (FileOutStream os = mFileSystem.createFile(uri,
+          CreateFilePOptions.newBuilder().setRecursive(true).build())) {
+        os.write(TEST_BYTES);
+      }
+    }
+    Assert.assertEquals(n, new File(mLocalUfsPath + "/direct").listFiles().length);
+    for (int i = 0; i < n; i++) {
+      checkCacheStatus(DIRECT_DIR + i, false, false);
+    }
+  }
+
+  @Test
+  public void writeNonDirect() throws Exception {
+    final int n = 3;
+    for (int i = 0; i < n; i++) {
+      AlluxioURI uri = new AlluxioURI(NON_DIRECT_DIR + i);
+      try (FileOutStream os = mFileSystem.createFile(uri,
+          CreateFilePOptions.newBuilder().setRecursive(true).build())) {
+        os.write(TEST_BYTES);
+      }
+    }
+    Assert.assertNull(new File(mLocalUfsPath + "/non_direct").listFiles());
+
+    for (int i = 0; i < n; i++) {
+      checkCacheStatus(NON_DIRECT_DIR + i, true, true);
+    }
+  }
+
+  private void checkCacheStatus(String path,
+      boolean shouldCacheBefore, boolean shouldCache) throws Exception {
+    AlluxioURI uri = new AlluxioURI(path);
+    Assert.assertEquals(shouldCacheBefore ? 100 : 0,
+        mFileSystem.getStatus(uri).getInMemoryPercentage());
+    try (FileInStream is = mFileSystem.openFile(uri)) {
+      IOUtils.copy(is, ByteStreams.nullOutputStream());
+    }
+    FileSystemUtils.waitForAlluxioPercentage(mFileSystem, uri, shouldCache ? 100 : 0);
```

Comment:
In case of `shouldCacheBefore == false` and `shouldCache == false`,
`waitForAlluxioPercentage` returns immediately when the expected cached percentage is 0. It takes time for this percentage to increase to above 0, so if the worker incorrectly caches the data which it wasn't supposed to cache, this test will fail to detect this.

Reply:
A better alternative is to look directly into the block master to see if the file has any associated blocks.